docs/tutorials/go-built-in.md

```diff
@@ -103,579 +101,527 @@ func NewKVStoreApplication() *KVStoreApplication {
+
+```go
+func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
+	return &KVStoreApplication{
+		db: db,
+	}
 }
 ```
 
-Don't worry if this does not compile yet.
-
-If the transaction does not have a form of `{bytes}={bytes}`, we return `1`
-code. When the same key=value already exist (same key and value), we return `2`
-code. For others, we return a zero code indicating that they are valid.
+Don't worry about `currentBatch` for now, we'll get to that later.
 
-Note that anything with non-zero code will be considered invalid (`-1`, `100`,
-etc.) by Tendermint Core.
+### 1.3.1 CheckTx
 
-Valid transactions will eventually be committed given they are not too big and
-have enough gas. To learn more about gas, check out ["the
-specification"](https://docs.tendermint.com/master/spec/abci/apps.html#gas).
+When Tendermint Core receives a new transaction, Tendermint asks the application
+if the transaction is acceptable. In our new application, let's implement some
+basic validation for the transactions it will receive.
 
-For the underlying key-value store we'll use
-[badger](https://github.com/dgraph-io/badger), which is an embeddable,
-persistent and fast key-value (KV) database.
+Add the following helper method to `app.go`:
 
 ```go
-import "github.com/dgraph-io/badger"
+func (app *KVStoreApplication) validateTx(tx []byte) uint32 {
+	parts := bytes.Split(tx, []byte("="))
+
+	// first, check that the transaction is not malformed.
+	if len(parts) != 2 {
+		return 1
+	}
+
+	key, value := parts[0], parts[1]
+
+	var txAlreadyExists bool
+
+	// ok, we have a well-formed transaction, next let's check that we haven't
+	// already stored the same transaction in the database before.
+	dbErr := app.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(key)
+		if err != nil {
+			if err != badger.ErrKeyNotFound {
+				return err
+			}
+			return nil
+		}
+		return item.Value(func(val []byte) error {
+			if bytes.Equal(val, value) {
+				txAlreadyExists = true
+			}
+			return nil
+		})
+	})
+
+	// There was an error processing the transaction.
+	// The application therefore cannot safely make progress so we crash.
+	if dbErr != nil {
+		panic(dbErr)
+	}
+
+	if txAlreadyExists {
+		return 2
+	}
+	return 0
 
-type KVStoreApplication struct {
- db           *badger.DB
- currentBatch *badger.Txn
 }
+```
 
-func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
- return &KVStoreApplication{
-  db: db,
- }
+And call it from within your `CheckTx` method:
+
+```go
+func (app *KVStoreApplication) CheckTx(req abcitypes.RequestCheckTx) abcitypes.ResponseCheckTx {
+	if code := app.validateTx(req.Tx); code != 0 {
+		return abcitypes.ResponseCheckTx{Code: code}
+	}
+	return abcitypes.ResponseCheckTx{Code: 0}
 }
 ```
 
+Any response with a non-zero code will be considered invalid by Tendermint. 
+Our `CheckTx` logic returns `0` to Tendermint when the transaction passes
+the validation checks.
```

Comment:
```suggestion
Our `CheckTx` logic returns `0` to Tendermint when a transaction passes
its validation checks.
```

It may also be worth reinforcing:

> Note that `CheckTx` does not _execute_ the transaction, it only verifies
> that the transaction _could_ be executed.  We don't know yet, at this point,
> whether the rest of the network has agreed to accept this transaction into
> a block.